<div>
<img src=https://www.institutedata.com/wp-content/uploads/2019/10/iod_h_tp_primary_c.svg width="300">
</div>

# Lab 3.1.4 
# *Python with MongoDB*

## Introduction to PyMongo

In [1]:
from IPython.core.display import display, HTML
import pymongo
from pymongo import MongoClient
print ('Mongo version ' + pymongo.__version__)

Mongo version 3.12.0


Start the mongod server (if it isn't lready running) by executing  
`mongod`  
at the command prompt. 

In [3]:
client = MongoClient('localhost', 27017)

In [6]:
# db = client.test
# collection = db.people
# collection.drop()

#Create a new database:

In [7]:
# mydb = client["mydatabase"]

In [25]:
import pymongo
from pymongo import MongoClient
import dns

print('pymongo version' + pymongo.__version__)


client = pymongo.MongoClient("mongodb+srv://Spring:chun1234@cluster0.iool0.mongodb.net/test_db?retryWrites=true&w=majority")



pymongo version3.12.0


In [26]:
print(client.list_database_names())

['test', 'test_db', 'admin', 'local']


In [36]:
db = client['test_db']
collection = db['test']
post = {'_id':0,'name':'Spring','gender':'male'}
collection.insert_one(post)

In [15]:
#ANSWER:
dblist = client.list_database_names()
if "test_db" in dblist:
    print("The database exists.")

The database exists.


Confirm that the database exists ... 

- list all databases in your system:

- check for the database by name:

In [28]:
mydb = client['test']

The new database was not found because mongodb is lazy: the db won't get created until data has been written to it!

Create a collection called "customers" (with object name `mycol`):

In [17]:
mycol=mydb['customers']

Create a document (i.e. a dictionary) with two name:value items 
("name" = "John", and "address" = "Highway 37") and insert 
it into the "customers" collection: 

In [18]:
#ANSWER:
mydict = { "name": "John", "address": "Highway 37" }
x = mycol.insert_one(mydict)

Now test for the existence of the database:

In [19]:
print(client.list_database_names())

['test', 'test_db', 'admin', 'local']


List all collections in the database:

In [29]:
print(mydb.list_collection_names())

['people', 'new_collection']


Insert another record in the "customers" collection 
("name" = "Peter", "address" = "Lowstreet 27") 
and return the value of the _id field:

In [20]:
#ANSWER:
mydict = { "name": "Peter", "address": "Lowstreet 27" }
x = mycol.insert_one(mydict)
print(x.inserted_id)

6104a14cdce0661c0558eb2f


Given the list of dicts below, insert multiple documents into 
the collection using the insert_many() method:

In [21]:
mylist = [
  { "name": "Amy", "address": "Apple st 652"},
  { "name": "Hannah", "address": "Mountain 21"},
  { "name": "Michael", "address": "Valley 345"},
  { "name": "Sandy", "address": "Ocean blvd 2"},
  { "name": "Betty", "address": "Green Grass 1"},
  { "name": "Richard", "address": "Sky st 331"},
  { "name": "Susan", "address": "One way 98"},
  { "name": "Vicky", "address": "Yellow Garden 2"},
  { "name": "Ben", "address": "Park Lane 38"},
  { "name": "William", "address": "Central st 954"},
  { "name": "Chuck", "address": "Main Road 989"},
  { "name": "Viola", "address": "Sideway 1633"}
]

In [22]:
x = mycol.insert_many(mylist)

Print a list of the _id values of the inserted documents:

In [23]:
#ANSWER:
print(x.inserted_ids)

[ObjectId('6104a154dce0661c0558eb30'), ObjectId('6104a154dce0661c0558eb31'), ObjectId('6104a154dce0661c0558eb32'), ObjectId('6104a154dce0661c0558eb33'), ObjectId('6104a154dce0661c0558eb34'), ObjectId('6104a154dce0661c0558eb35'), ObjectId('6104a154dce0661c0558eb36'), ObjectId('6104a154dce0661c0558eb37'), ObjectId('6104a154dce0661c0558eb38'), ObjectId('6104a154dce0661c0558eb39'), ObjectId('6104a154dce0661c0558eb3a'), ObjectId('6104a154dce0661c0558eb3b')]


Execute the next cell to insert a list of dicts with specified `_id`s:

In [48]:
mylist = [
  { "_id": 1, "name": "John", "address": "Highway 37"},
  { "_id": 2, "name": "Peter", "address": "Lowstreet 27"},
  { "_id": 3, "name": "Amy", "address": "Apple st 652"},
  { "_id": 4, "name": "Hannah", "address": "Mountain 21"},
  { "_id": 5, "name": "Michael", "address": "Valley 345"},
  { "_id": 6, "name": "Sandy", "address": "Ocean blvd 2"},
  { "_id": 7, "name": "Betty", "address": "Green Grass 1"},
  { "_id": 8, "name": "Richard", "address": "Sky st 331"},
  { "_id": 9, "name": "Susan", "address": "One way 98"},
  { "_id": 10, "name": "Vicky", "address": "Yellow Garden 2"},
  { "_id": 11, "name": "Ben", "address": "Park Lane 38"},
  { "_id": 12, "name": "William", "address": "Central st 954"},
  { "_id": 13, "name": "Chuck", "address": "Main Road 989"},
  { "_id": 14, "name": "Viola", "address": "Sideway 1633"}
]
x = mycol.insert_many(mylist)
print(x.inserted_ids)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]


Now try inserting a new dict with an existing `_id`:

In [51]:
# x = mycol.insert_one({ "_id": 14, "name": "Manuel", "address": "Barcelona"})

So, if we want to manage `_id`s in code, we need to be careful!

This returns the first document in the collection:

In [52]:
x = mycol.find_one()
print(x)

{'_id': ObjectId('61014bf2d74ed26e3153a3a3'), 'name': 'John', 'address': 'Highway 37'}


Do the same for the document containing "name" = "Hannah":

In [53]:
x = mycol.find_one({"name": "Hannah"})
print(x)

{'_id': ObjectId('61014ca3d74ed26e3153a3a6'), 'name': 'Hannah', 'address': 'Mountain 21'}


This returns (and prints) all documents in the collection:

In [54]:
for x in mycol.find():
  print(x)

{'_id': ObjectId('61014bf2d74ed26e3153a3a3'), 'name': 'John', 'address': 'Highway 37'}
{'_id': ObjectId('61014c67d74ed26e3153a3a4'), 'name': 'Peter', 'address': 'Lowstreet 27'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a5'), 'name': 'Amy', 'address': 'Apple st 652'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a6'), 'name': 'Hannah', 'address': 'Mountain 21'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a7'), 'name': 'Michael', 'address': 'Valley 345'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a8'), 'name': 'Sandy', 'address': 'Ocean blvd 2'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a9'), 'name': 'Betty', 'address': 'Green Grass 1'}
{'_id': ObjectId('61014ca3d74ed26e3153a3aa'), 'name': 'Richard', 'address': 'Sky st 331'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ab'), 'name': 'Susan', 'address': 'One way 98'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ac'), 'name': 'Vicky', 'address': 'Yellow Garden 2'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ad'), 'name': 'Ben', 'address': 'Park Lane 38'}
{'_id': ObjectI

This returns only the name and address fields:

In [65]:
for x in mycol.find({},{ "_id": 0, "name": 1, "address": 1 }):
  print(x)

{'name': 'John', 'address': 'Highway 37'}
{'name': 'Peter', 'address': 'Lowstreet 27'}
{'name': 'Amy', 'address': 'Apple st 652'}
{'name': 'Hannah', 'address': 'Mountain 21'}
{'name': 'Michael', 'address': 'Valley 345'}
{'name': 'Sandy', 'address': 'Ocean blvd 2'}
{'name': 'Betty', 'address': 'Green Grass 1'}
{'name': 'Richard', 'address': 'Sky st 331'}
{'name': 'Susan', 'address': 'One way 98'}
{'name': 'Vicky', 'address': 'Yellow Garden 2'}
{'name': 'Ben', 'address': 'Park Lane 38'}
{'name': 'William', 'address': 'Central st 954'}
{'name': 'Chuck', 'address': 'Main Road 989'}
{'name': 'Viola', 'address': 'Sideway 1633'}
{'name': 'John', 'address': 'Highway 37'}
{'name': 'Peter', 'address': 'Lowstreet 27'}
{'name': 'Amy', 'address': 'Apple st 652'}
{'name': 'Hannah', 'address': 'Mountain 21'}
{'name': 'Michael', 'address': 'Valley 345'}
{'name': 'Sandy', 'address': 'Ocean blvd 2'}
{'name': 'Betty', 'address': 'Green Grass 1'}
{'name': 'Richard', 'address': 'Sky st 331'}
{'name': 'Susa

Print only the `_id` and name fields:

In [27]:
for x in mycol.find({},{ "_id": 0, "name": 1,"address":1 }):
  print(x)

{'name': 'John', 'address': 'Highway 37'}
{'name': 'Peter', 'address': 'Lowstreet 27'}
{'name': 'Amy', 'address': 'Apple st 652'}
{'name': 'Sandy', 'address': 'Ocean blvd 2'}
{'name': 'Vicky', 'address': 'Yellow Garden 2'}
{'name': 'Chuck', 'address': 'Main Road 989'}
{'name': 'Betty', 'address': 'Green Grass 1'}
{'name': 'Susan', 'address': 'One way 98'}
{'name': 'Ben', 'address': 'Park Lane 38'}
{'name': 'Michael', 'address': 'Canyon 123'}
{'name': 'Richard', 'address': 'Sky st 331'}
{'name': 'Viola', 'address': 'Sideway 1633'}
{'name': 'William', 'address': 'Central st 954'}
{'name': 'John', 'address': 'Highway 37'}
{'name': 'Peter', 'address': 'Lowstreet 27'}
{'name': 'Amy', 'address': 'Apple st 652'}
{'name': 'Hannah', 'address': 'Mountain 21'}
{'name': 'Michael', 'address': 'Valley 345'}
{'name': 'Sandy', 'address': 'Ocean blvd 2'}
{'name': 'Betty', 'address': 'Green Grass 1'}
{'name': 'Richard', 'address': 'Sky st 331'}
{'name': 'Susan', 'address': 'One way 98'}
{'name': 'Vicky'

So, we must explicitly use `"_id": 0` to exclude it, but for other fields we simply omit them from the dict argument.

To include field conditionals in a query, we use `$` operators. This finds addresses starting with "S" or greater:

In [32]:
myquery = { "address": { "$gt": "Q" } }
mydoc = mycol.find(myquery)
for x in mydoc:
  print(x)

{'_id': ObjectId('61014ca3d74ed26e3153a3ac'), 'name': 'Vicky', 'address': 'Yellow Garden 2'}
{'_id': ObjectId('61014ca3d74ed26e3153a3aa'), 'name': 'Richard', 'address': 'Sky st 331'}
{'_id': ObjectId('61014ca3d74ed26e3153a3b0'), 'name': 'Viola', 'address': 'Sideway 1633'}
{'_id': ObjectId('6104a154dce0661c0558eb32'), 'name': 'Michael', 'address': 'Valley 345'}
{'_id': ObjectId('6104a154dce0661c0558eb35'), 'name': 'Richard', 'address': 'Sky st 331'}
{'_id': ObjectId('6104a154dce0661c0558eb37'), 'name': 'Vicky', 'address': 'Yellow Garden 2'}
{'_id': ObjectId('6104a154dce0661c0558eb3b'), 'name': 'Viola', 'address': 'Sideway 1633'}


Here are some more comparison operators:

            $gt, $gte, $eq, $in, $nin, $exists, $and, $or, $not
            
Experiment with these until you understand how to use them.

Now find all docs with an address that begins with "S":  
(HINT: The value for "address" in the argument should be the regex-based dict { "$regex": "^S" }.)

In [37]:
#ANSWER:
myquery = { "address": { "$regex": "L" } }
mydoc = mycol.find(myquery)
for x in mydoc:
  print(x)

{'_id': ObjectId('61014c67d74ed26e3153a3a4'), 'name': 'Peter', 'address': 'Lowstreet 27'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ad'), 'name': 'Ben', 'address': 'Park Lane 38'}
{'_id': ObjectId('6104a14cdce0661c0558eb2f'), 'name': 'Peter', 'address': 'Lowstreet 27'}
{'_id': ObjectId('6104a154dce0661c0558eb38'), 'name': 'Ben', 'address': 'Park Lane 38'}


Sorting can be applied by invoking the Sort() method after the Find() method. Sort the collection by the name field:

In [71]:
#ANSWER:
mydoc = mycol.find().sort("name")
for x in mydoc:
  print(x)

{'_id': ObjectId('61014ca3d74ed26e3153a3a5'), 'name': 'Amy', 'address': 'Apple st 652'}
{'_id': 3, 'name': 'Amy', 'address': 'Apple st 652'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ad'), 'name': 'Ben', 'address': 'Park Lane 38'}
{'_id': 11, 'name': 'Ben', 'address': 'Park Lane 38'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a9'), 'name': 'Betty', 'address': 'Green Grass 1'}
{'_id': 7, 'name': 'Betty', 'address': 'Green Grass 1'}
{'_id': ObjectId('61014ca3d74ed26e3153a3af'), 'name': 'Chuck', 'address': 'Main Road 989'}
{'_id': 13, 'name': 'Chuck', 'address': 'Main Road 989'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a6'), 'name': 'Hannah', 'address': 'Mountain 21'}
{'_id': 4, 'name': 'Hannah', 'address': 'Mountain 21'}
{'_id': ObjectId('61014bf2d74ed26e3153a3a3'), 'name': 'John', 'address': 'Highway 37'}
{'_id': 1, 'name': 'John', 'address': 'Highway 37'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a7'), 'name': 'Michael', 'address': 'Valley 345'}
{'_id': 5, 'name': 'Michael', 'address': 'Valley

Now sort in reverse order (HINT: The Sort() method takes an optional second parameter.)

In [72]:
#ANSWER:
mydoc = mycol.find().sort("name", direction=pymongo.DESCENDING)
for x in mydoc:
  print(x)

{'_id': ObjectId('61014ca3d74ed26e3153a3ae'), 'name': 'William', 'address': 'Central st 954'}
{'_id': 12, 'name': 'William', 'address': 'Central st 954'}
{'_id': ObjectId('61014ca3d74ed26e3153a3b0'), 'name': 'Viola', 'address': 'Sideway 1633'}
{'_id': 14, 'name': 'Viola', 'address': 'Sideway 1633'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ac'), 'name': 'Vicky', 'address': 'Yellow Garden 2'}
{'_id': 10, 'name': 'Vicky', 'address': 'Yellow Garden 2'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ab'), 'name': 'Susan', 'address': 'One way 98'}
{'_id': 9, 'name': 'Susan', 'address': 'One way 98'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a8'), 'name': 'Sandy', 'address': 'Ocean blvd 2'}
{'_id': 6, 'name': 'Sandy', 'address': 'Ocean blvd 2'}
{'_id': ObjectId('61014ca3d74ed26e3153a3aa'), 'name': 'Richard', 'address': 'Sky st 331'}
{'_id': 8, 'name': 'Richard', 'address': 'Sky st 331'}
{'_id': ObjectId('61014c67d74ed26e3153a3a4'), 'name': 'Peter', 'address': 'Lowstreet 27'}
{'_id': 2, 'name': 'Peter',

A single record can be deleted by specifying some criterion:

In [38]:
mycol.delete_one({ "address": "Mountain 21" })

Now delete all docs with the 2-digit `Id` values:

In [39]:
#ANSWER:
mycol.delete_many({ "_id": {"$lt": 15} })
for x in mycol.find():
  print(x)

{'_id': ObjectId('61014bf2d74ed26e3153a3a3'), 'name': 'John', 'address': 'Highway 37'}
{'_id': ObjectId('61014c67d74ed26e3153a3a4'), 'name': 'Peter', 'address': 'Lowstreet 27'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a5'), 'name': 'Amy', 'address': 'Apple st 652'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a8'), 'name': 'Sandy', 'address': 'Ocean blvd 2'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ac'), 'name': 'Vicky', 'address': 'Yellow Garden 2'}
{'_id': ObjectId('61014ca3d74ed26e3153a3af'), 'name': 'Chuck', 'address': 'Main Road 989'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a9'), 'name': 'Betty', 'address': 'Green Grass 1'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ab'), 'name': 'Susan', 'address': 'One way 98'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ad'), 'name': 'Ben', 'address': 'Park Lane 38'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a7'), 'name': 'Michael', 'address': 'Canyon 123'}
{'_id': ObjectId('61014ca3d74ed26e3153a3aa'), 'name': 'Richard', 'address': 'Sky st 331'}
{'_id': Object

This would delete all docs:
`x = mycol.delete_many({})`

This would remove the collection:
`mycol.drop()`

Change the first instance of "address" == "Valley 345" to "Canyon 123" using update_one().  
(HINT: The 1st paramater of update_one() is the criterion (query); the 2nd is dict specifying the field to change and its new value.) 

In [42]:
#ANSWER:
myquery = { "address": "Apple st 652" }
newvalues = { "$set": { "address": "Apple st 6521" } }
mycol.update_one(myquery, newvalues)
for x in mycol.find().limit(5):
  print(x)

{'_id': ObjectId('61014bf2d74ed26e3153a3a3'), 'name': 'John', 'address': 'Highway 37'}
{'_id': ObjectId('61014c67d74ed26e3153a3a4'), 'name': 'Peter', 'address': 'Lowstreet 27'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a5'), 'name': 'Amy', 'address': 'Apple st 6521'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a8'), 'name': 'Sandy', 'address': 'Ocean blvd 2'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ac'), 'name': 'Vicky', 'address': 'Yellow Garden 2'}


The limit() method can be applied after the find() method to limit the number of docs returned. Show the first 5 docs:

In [91]:
#ANSWER:
myresult = mycol.find()
for x in myresult:
  print(x)

{'_id': ObjectId('61014bf2d74ed26e3153a3a3'), 'name': 'John', 'address': 'Highway 37'}
{'_id': ObjectId('61014c67d74ed26e3153a3a4'), 'name': 'Peter', 'address': 'Lowstreet 27'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a5'), 'name': 'Amy', 'address': 'Apple st 652'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a7'), 'name': 'Michael', 'address': 'Canyon 123'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a8'), 'name': 'Sandy', 'address': 'Ocean blvd 2'}
{'_id': ObjectId('61014ca3d74ed26e3153a3a9'), 'name': 'Betty', 'address': 'Green Grass 1'}
{'_id': ObjectId('61014ca3d74ed26e3153a3aa'), 'name': 'Richard', 'address': 'Sky st 331'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ab'), 'name': 'Susan', 'address': 'One way 98'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ac'), 'name': 'Vicky', 'address': 'Yellow Garden 2'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ad'), 'name': 'Ben', 'address': 'Park Lane 38'}
{'_id': ObjectId('61014ca3d74ed26e3153a3ae'), 'name': 'William', 'address': 'Central st 954'}
{'_id': Obj

In [90]:
post_count = mycol.count_documents({})
print(post_count)

13


## PyMongo for Data Science

MongoDB has many more features of interest to developers, but the main focus of a data scientist will be wrangling and munging the data. It may or may not be desirable to do all the data munging in Pandas; for a large, distributed database, it may be imperative to perform aggregation in MongoDB. 

In [2]:
# Ref:  https://rsandstroem.github.io/MongoDBDemo.html

# import os
# import pandas as pd
# import numpy as np
# from IPython.core.display import display, HTML
# import pymongo
# from pymongo import MongoClient
# print ('Mongo version', pymongo.__version__)
# client = MongoClient('localhost', 27017)
# db = client.test
# collection = db.people

Mongo version 3.12.0


This code creates a database named "test" and populates it from a JSON file using the mongoimport program (executed in the operating systems itself, rather than in Python):

In [39]:
import pandas as pd
import pymongo
from pymongo import MongoClient
import dns
import json

print('pymongo version' + pymongo.__version__)

client = pymongo.MongoClient("mongodb+srv://Spring:chun1234@cluster0.iool0.mongodb.net/mongo_db?retryWrites=true&w=majority")
db = client.test
collection = db['people']

#collection.insert_one({'_id':50,'name':'Sally'})

pymongo version3.12.0


In [16]:
pwd()

'C:\\Users\\chun.dai\\OneDrive - Aftermarket Network Australia\\Documents\\Institute_of_data\\Module_3'

In [38]:
cursor = collection.find().sort('Age',pymongo.ASCENDING).limit(3)
for doc in cursor:
    print doc

SyntaxError: Missing parentheses in call to 'print'. Did you mean print(doc)? (<ipython-input-38-489dfd210fc9>, line 3)

In [35]:
import os
os.system('mongoimport --uri mongodb+srv://Spring:chun1234@cluster0.iool0.mongodb.net/test --collection people --type json --file "C:/Users/chun.dai/OneDrive - Aftermarket Network Australia/Documents/Institute_of_data/DATA/dummyData.json"')

0

In [83]:
os.chdir("..")

In [84]:
pwd()

'C:\\Users\\chun.dai\\OneDrive - Aftermarket Network Australia\\Documents\\Institute_of_data'

In [36]:
db = client.test
collection = db.people
cursor = collection.find().sort('Age',pymongo.ASCENDING).limit(3)
for doc in cursor:
    print (doc)

{'_id': ObjectId('610cdd45ddb9050e840129e0'), 'Name': 'Sawyer, Neve M.', 'Age': 18, 'Country': 'Serbia', 'Location': '-34.37446, 174.0838'}
{'_id': ObjectId('610cdd45ddb9050e84012988'), 'Name': 'Townsend, Cadman I.', 'Age': 19, 'Country': 'Somalia', 'Location': '-87.69188, -144.16138'}
{'_id': ObjectId('610cdd45ddb9050e840129d6'), 'Name': 'Shepard, Peter R.', 'Age': 20, 'Country': 'Saint Helena, Ascension and Tristan da Cunha', 'Location': '28.90219, 66.86083'}


Here is a small demonstration of the MongoDB aggregation framework. We want to create a table of the number of persons in each country and their average age. To do it we group by country. We extract the results from MongoDB aggregation into a pandas dataframe, and use the country as index.

In [40]:
pipeline = [
        {"$group": {"_id":"$Country",
             "AvgAge":{"$avg":"$Age"},
             "Count":{"$sum":1},
        }},
        {"$sort":{"Count":-1,"AvgAge":1}}
]
aggResult = collection.aggregate(pipeline) # returns a cursor

df1 = pd.DataFrame(list(aggResult)) # use list to turn the cursor to an array of documents
df1 = df1.set_index("_id")
df1.head()

AvgAge  Count
_id                                                
China                              46.250000      4
Antarctica                         46.333333      3
Guernsey                           48.333333      3
Puerto Rico                        26.500000      2
Heard Island and Mcdonald Islands  29.000000      2

For simple cases one can either use a cursor through find("search term") or use the "$match" operator in the aggregation framework, like this:

In [41]:
pipeline = [
        {"$match": {"Country":"China"}},
]
aggResult = collection.aggregate(pipeline)
df2 = pd.DataFrame(list(aggResult))
df2.head()

_id                Name  Age Country  \
0  610cdd45ddb9050e84012985    Holman, Hasad O.   32   China   
1  610cdd45ddb9050e840129c4      Byrd, Dante A.   43   China   
2  610cdd45ddb9050e840129ca  Carney, Tamekah I.   57   China   
3  610cdd45ddb9050e840129d8    Mayer, Violet U.   53   China   

            Location  
0  39.9127, 116.3833  
1        31.2, 121.5  
2    45.75, 126.6333  
3             40, 95

Now we can apply all the power of Python libraries to analyse and visualise the data. Here, we will use the folium package to plot markers for the locations of the people we just found in China (click on a marker to see their data):

In [ ]:
# Un-comment and execute to install folium pkg (1st time only):
# import sys
# !{sys.executable} -m pip install folium

In [42]:
import folium
print ('Folium version ' + folium.__version__)

world_map = folium.Map(location = [35, 100], zoom_start = 4)
for i in range(len(df2)):
    location = [float(loc) for loc in df2.Location[i].split(',')]
    folium.Marker(location = location, popup = df2.Name[i] + ', age:' + str(df2.Age[i])).add_to(world_map)
    
world_map

Folium version unknown


## HOMEWORK:


1. Read up on how to perform aggregation in mongoDB. Insert a duplicate record into the collection:
        mydict = { "name": "John", "address": "Highway 37" }
   Now write a command to find docs with a duplicate "name" field (using aggregation) and remove them.  
   Print the collection.

2. Read up on how to apply indexes in mongoDB. Create an index on the "name" and "address" fields in this collection.
   Print the indexes for the collection.



> 
>
>






---



---



> > > > > > > > > © 2021 Institute of Data


---



---



